doc/tutorials/debug.rst

```diff
@@ -0,0 +1,371 @@
+.. _acrn-debug:
+
+ACRN Debugging Tools
+####################
+
+This document describes how to use ACRN tools to collect log
+and trace data for debugging.
+
+ACRN Console Command
+********************
+
+The ACRN console provides shell commands for a user to check system states
+and environment settings.
+
+Help
+====
+
+"help" provides information about the supported hypervisor shell commands:
+
+`help`_
+   Displays information about supported hypervisor shell commands
+
+`vm_list`_
+   Lists all VMs, displaying VM Name, VM ID, and VM State (ON=running)
+
+`vcpu_list`_
+   Lists all VCPUs in all VMs
+
+`vcpu_dumpreg`_ <vm_id, vcpu_id>``
+   Dumps registers for a specific vCPU
+
+`dumpmem`_ <hva, length>``
+   Dumps host memory, starting a given address, and for a given length (in bytes)
+
+sos_console
+   Switches to the SOS's console
+
+`int`_
+   Lists interrupt information per CPU
+
+`pt`_
+   Shows pass-through device information
+
+`dump_ioapic`_
+   Shows native ioapic information
+
+vmexit
+   Shows vmexit profiling
+
+loglevel [console_loglevel] [mem_loglevel]
+   Get (when no parameters are given) or Set loglevel [0 (none) - 6 (verbose)]
+   for the console and optionally for memory
+
+`cpuid`_ <leaf> [subleaf]
+   Displays the CPUID leaf [subleaf], in hexadecimal
+
+`reboot`
+   Triggers a system reboot (immediately)
+
+vm_list
+=======
+
+"vm_list" provides the name of each virtual machine and its corresponding ID and
+state.
+
+.. figure:: images/debug_image8.png
+   :align: center
+
+   vm_list information
+
+vcpu_list
+=========
+
+"vcpu_list" provides information about virtual CPUs (vCPU), including the VM ID, PCPU
+ID, VCPU ID, VCPU ROLE (primary or secondary), and VCPU STATE (running
+or paused).
+
+.. figure:: images/debug_image7.png
+   :align: center
+
+   vcpu_list information
+
+vcpu_dumpreg
+============
+
+"vcpu_dumpreg vmid cpuid" provides vCPU related information such as
+registers values, etc.
+
+In the following example, we dump vCPU0 RIP register value and get into
+the SOS to search for the currently running function, using these
+commands::
+
+   cat /proc/kallsyms | grep RIP_value
+
+As you can see, vCPU0 is running in
+function ``acpi_idle_do_entry``.
+
+.. figure:: images/debug_image10.png
+   :align: center
+
+   vcpu_dumpreg information
+
+.. figure:: images/debug_image9.png
+   :align: center
+
+   system map information
+
+dumpmem
+=======
+
+"dumpmem mem_address" provides the specified memory target data such as
+the physical CPU (pCPU) number, etc.
+
+In this example, we know the pCPU active bitmap and physical CPU number
+physical memory address through
+``out/hv/apl-mrb-sbl/hypervisor/acrn.map``.
```

Comment:
This path is very specific to how you built the hypervisor. Do we provide that `*.map` file with the `service-os` bundle (or any other bundle?